### PR TITLE
SocketAsyncContext.Unix: remove Lock from IsReady

### PR DESCRIPTION
**Pull request by tmds**
triggered by https://github.com/dotnet/runtime/pull/36214

@adamsitnik @stephentoub ptal

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 1]
```diff
@@ -1100,6 +1102,7 @@ public void CancelAndContinueProcessing(TOperation op)
                                 if (_tail == null)
                                 {
                                     _state = QueueState.Ready;
+                                    _sequenceNumber++;
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:1105]: This change is unrelated. All code paths that set _state to QueueState.Ready increment _sequenceNumber. I'm not sure why it is not done at this location, so I've added it.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 2]
```diff
@@ -742,22 +742,24 @@ public void Init()
...
+            // This ensures the sequence number is read, before executing the operation.
+            // When the operation can't complete, StartAsyncOperation compares against the current
+            // sequence number to see if the operation should be tried again.
+            public bool IsReady(SocketAsyncContext context, out int? observedSequenceNumber)
             {
-                using (Lock())
-                {
-                    observedSequenceNumber = _sequenceNumber;
-                    bool isReady = (_state == QueueState.Ready) || (_state == QueueState.Stopped);
+                QueueState state = (QueueState)Volatile.Read(ref Unsafe.As<QueueState, int>(ref _state));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:752]: QueueState is a byte in size.  This doesn't look safe / correct.
```C#
private enum QueueState : byte
```

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:752]: Oops. I've changed it to `int`.

**Comment by tmds** [comment]
@adamsitnik I hope to see at least the 3k rps you measured in https://github.com/dotnet/runtime/pull/36214. And I hope for some more because this is now used for sends and receives.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 3]
```diff
@@ -742,22 +742,24 @@ public void Init()
...
-            // IsReady returns the current _sequenceNumber, which must be passed to StartAsyncOperation below.
-            public bool IsReady(SocketAsyncContext context, out int observedSequenceNumber)
+            // IsReady returns whether an operation can be executed immediately without queueing it.
+            // When IsReady returns true, the observedSequenceNumber is set.
+            // This ensures the sequence number is read, before executing the operation.
+            // When the operation can't complete, StartAsyncOperation compares against the current
+            // sequence number to see if the operation should be tried again.
+            public bool IsReady(SocketAsyncContext context, out int? observedSequenceNumber)
             {
-                using (Lock())
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:748]: What's the rationale for why it's ok to not take the lock?

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:748]: The return value is still based on `_state`. Doing `Volatile.Read` will give the same result.

The role of the `observedSequenceNumber` is to ensure we don't miss events that require retrying.

By doing `Volatile.Read` before the op (which can't re-order), we can compare in `StartAsyncOperation`.

Without the `Lock` it is possible we observe the sequence number increment before state becomes ready. In that case `observedSequenceNumber` will be `null`, and `StartAsyncOperation` will retry the operation.

**Comment by tmds** [comment]
@adamsitnik for benchmarking: [System.Net.Sockets.dll.tar.gz](https://github.com/dotnet/runtime/files/4652697/System.Net.Sockets.dll.tar.gz)

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:748]: Thanks.  Can you include the relevant details in a comment?

**Comment by adamsitnik** [comment]
The CI fails with following error:

```log
System/Net/Sockets/SocketAsyncContext.Unix.cs(1571,49): error CS1503: (NETCORE_ENGINEERING_TELEMETRY=Build) Argument 2: cannot convert from 'out int' to 'out int?'
```

**Comment by tmds** [comment]
@adamsitnik not sure if you saw the change to make CI pass? Here is the dll: [System.Net.Sockets.dll.tar.gz](https://github.com/dotnet/runtime/files/4677060/System.Net.Sockets.dll.tar.gz)

**Comment by adamsitnik** [comment]
@tmds the results
obraz

I've uploaded jsonplatform trace to `traces/36705_isready_nolock`

**Comment by tmds** [comment]
Thanks for benchmarking Adam. For Citrine JSON platform, the improvements seems to be round the expected 3k (0.24%). Notable better improvement on ARM 32-core: 3.28%.

About the measured regressions, it doesn't make sense why a lock (2 Interlocked) can perform better than two Volatile.Reads to me.

I'm in for this change because the complexity is not so high. The tiny improvement applies to every socket operation.

**Comment by adamsitnik** [comment]
> About the measured regressions,

If you are talking about the `mono` machine JSON results I think that the results produced on this machine are always quite unstable and you can ignore it

**Comment by tmds** [comment]
Is this good to merge?

**Comment by stephentoub** [comment]
> Is this good to merge?

In Adam's results, there are a bunch of regressions (not just on the mono machines) that are larger than the improvement in json on citrine.  That's not a concern?

**Comment by tmds** [comment]
> In Adam's results, there are a bunch of regressions (not just on the mono machines) that are larger than the improvement in json on citrine. That's not a concern?

From the benchmark results it doesn't look like it's an every scenario wins change. Though I don't understand why two volatile reads wouldn't always be cheaper than two compareexchanges.
Per Adam's comment we can ignore the results for mono. I think overall the trends looks good then.

**Comment by stephentoub** [comment]
> I think overall the trends looks good then.

I'm struggling to see that.  Can you elaborate?  There was a notable increase in one benchmark on ARM, but two of the three shown regressed on AMD, two of the three regressed on Perf 12 cores, and while there was a measurable improvement in Fortunes Platform on citrine, there was an almost commensurate regression in Fortunes Batching on citrine.

Is it possible this was just a bad run of the metrics?

While I agree in theory removing a lock should be helpful, at the same time lock-freedom generally makes code more complex and harder to reason about and maintain, and so I'd only want to take such a change if we can prove to ourselves it really moves things in the right direction meaningfully.

**Comment by tmds** [comment]
@stephentoub I'm basing myself mostly on ARM results and time spent in traces inside `IsReady`. I'm also making some assumptions about the cost of volatileread vs compareexchange, which are maybe incorrect.

@adamsitnik what do you think about these regressions? To we have an idea on how much the same benchmark varies across successive runs? Are we confident these are regressions?

I'm fine if you want to close this because the gain (0.24% on jsonplatform) is not worth the investigation. Though the gain on ARM seems worthwhile. And it will apply to any socket operation, not just TE 100% load cases.

**Comment by stephentoub** [comment]
> I'm fine if you want to close this

I don't; I want it to be successful :)  I just want to make sure we're only taking it when we can prove that it actually is.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 4]
```diff
@@ -753,22 +753,27 @@ public void Init()
                 _sequenceNumber = 0;
             }
 
-            // IsReady returns the current _sequenceNumber, which must be passed to StartAsyncOperation below.
-            public bool IsReady(SocketAsyncContext context, out int observedSequenceNumber)
+            // IsReady returns whether an operation can be executed immediately.
+            // observedSequenceNumber must be passed to StartAsyncOperation.
+            public bool IsReady(SocketAsyncContext context, out int? observedSequenceNumber)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:758]: Can we keep `observedSequenceNumber` as a non-nullable int? It's only used when IsReady returns true, so we should be able to just return 0 or whatever when that's not the case. Seems like a non-nullable int would be slightly more efficient here.

**Comment by benaadams** [comment]
`.IsReady` does currently show up under `JIT_MonExit_Portable` in the traces

image

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 4 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:758]: > it's only used when IsReady returns true

When `IsReady` returns `false`, the `null` is used to ensure that `StartAsyncOperation` will try the operation in case the queue becomes ready in the meanwhile. To not use a nullable int here, the best value to use is `_sequenceNumber - 1` (when returning `false`), because that will fail the sequence number equals check in `StartAsyncOperation`.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 5]
```diff
@@ -753,18 +753,29 @@ public void Init()
...
             {
-                using (Lock())
+                // It is safe to read _state and _sequence without using Lock.
+                // - The return value is soly based on Volatile.Read of _state.
+                // - The Volatile.Read of _sequenceNumber ensures we read a value before executing the operation.
+                //   This is needed to retry the operation in StartAsyncOperation in case the _sequenceNumber incremented.
+                // - Because no Lock is taken, it is possible we observe a sequence number increment before the state
+                //   becomes Ready. When that happens, observedSequenceNumber is decremented, and StartAsyncOperation will
+                //   execute the operation because the sequence number won't match.
+                QueueState state = (QueueState)Volatile.Read(ref Unsafe.As<QueueState, int>(ref _state));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:767]: Nit: could you add a `Debug.Assert(sizeof(QueueState) == sizeof(int));` or something along those lines?  I'm just imagining a year from now one of us comes along and says "aha, QueueState could be a byte instead of an int to save space" and fails to notice this silent requirement that they're the same size.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 6]
```diff
@@ -753,18 +753,29 @@ public void Init()
...
+                observedSequenceNumber = Volatile.Read(ref _sequenceNumber);
+
+                bool isReady = state == QueueState.Ready || state == QueueState.Stopped;
+                if (!isReady)
                 {
-                    observedSequenceNumber = _sequenceNumber;
-                    bool isReady = (_state == QueueState.Ready) || (_state == QueueState.Stopped);
+                    observedSequenceNumber--;
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:773]: Is there an enforced cap on the sequence number somewhere?  I'm wondering what happens if, for example, this wraps around such that observedSequenceNumber is int.MinValue and observedSequenceNumber-- is int.MaxValue.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 7]
```diff
@@ -753,18 +753,29 @@ public void Init()
...
             }
 
-            // IsReady returns the current _sequenceNumber, which must be passed to StartAsyncOperation below.
+            // IsReady returns whether an operation can be executed immediately.
+            // observedSequenceNumber must be passed to StartAsyncOperation.
             public bool IsReady(SocketAsyncContext context, out int observedSequenceNumber)
             {
-                using (Lock())
+                // It is safe to read _state and _sequence without using Lock.
+                // - The return value is soly based on Volatile.Read of _state.
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:761]:
```suggestion
                // - The return value is soley based on Volatile.Read of _state.
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 6 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:773]: There is no cap on sequence number. The observed sequence number gets checked for equality in `StartAsyncOperation`. It is a wrapping counter for the nr of times the queue becomes ready. Decrementing it is the same as pretending we have 'observed' the previous time the queue became ready, so `StartAsyncOperation` considers `IsReady` state as a new thing.
I think there is nothing special at `int.{Min,Max}Value` because we're doing `++`, `--` and `!=` operations only.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 5 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:769]: :+1: I've been wanting to make this more explicit.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:775]: Ok, I wasn't sure if it was ever compared with `<`.  If it's only ever compared with `==` seems fine.  Out of curiosity, what would happen if 4B operations occurred between the time we read the value and compare it such that it fully wrapped and was again the same?  I assume if that's a (super rare) problem, it's already a problem.

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:775]: > Out of curiosity, what would happen if 4B operations occurred between the time we read the value and compare it such that it fully wrapped and was again the same?

We'd loose an event, and the socket operation will never complete. Since there is a limited nr of cases that cause increments, we don't need to worry about reaching 4B increments.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:775]: Again, not actually concerned, but trying to understand... re "Since there is a limited nr of cases that cause increments, we don't need to worry about reaching 4B increments", what would prevent one thread from stalling in between the get and the check, and other threads processing 4B operations in between?

**Review by stephentoub — APPROVED** [review]
